smart_open/s3.py

```diff
@@ -279,47 +279,98 @@ class _SeekableRawReader(object):
         self._object_kwargs = object_kwargs if object_kwargs else {}
 
-    def seek(self, position):
-        """Seek to the specified position (byte offset) in the S3 key.
+    def seek(self, offset, whence=constants.WHENCE_START):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
 
-        :param int position: The byte offset from the beginning of the key.
+        Returns the position after seeking.
         """
+        if whence not in constants.WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % constants.WHENCE_CHOICES)
+
         #
         # Close old body explicitly.
         # When first seek() after __init__(), self._body is not exist.
         #
         if self._body is not None:
             self._body.close()
         self._body = None
-        self._position = position
 
-    def _load_body(self):
-        """Build a continuous connection with the remote peer starts from the current postion.
-        """
-        range_string = smart_open.utils.make_range_string(self._position)
-        logger.debug('content_length: %r range_string: %r', self._content_length, range_string)
+        start = None
+        stop = None
+        if whence == constants.WHENCE_START:
+            start = max(0, offset)
+        elif whence == constants.WHENCE_CURRENT:
+            start = max(0, offset + self._position)
+        else:
+            stop = max(0, -offset)
 
-        if self._position == self._content_length == 0 or self._position == self._content_length:
-            #
-            # When reading, we can't seek to the first byte of an empty file.
-            # Similarly, we can't seek past the last byte.  Do nothing here.
-            #
+        if self._content_length is not None and \
+                ((start is not None and start >= self._content_length) or stop == 0):
+            # avoid calling _load_body() if we are at the end of the file
             self._body = io.BytesIO()
+            self._position = self._content_length
         else:
-            self._body = _get(
+            range_string = smart_open.utils.make_range_string(start, stop)
+            self._load_body(range_string)  # sets self._position
+
+        return self._position
+
+    def _load_body(self, range_string=None):
+        """Build a continuous connection with the remote peer starts from the current postion.
+        """
+        if range_string is None:
+            range_string = smart_open.utils.make_range_string(self._position)
+        logger.debug('range_string: %r', range_string)
+
+        try:
+            # Optimistically try to fetch the requested content range.
+            response = _get(
                 self._object,
                 version=self._version_id,
                 Range=range_string,
                 **self._object_kwargs
-            )['Body']
+            )
+        except IOError as error:
```

Comment:
This exception-handling block is a bit too busy. I think it's worth simplifying it. Make it obvious what you're doing and why.

To me, it looks like there are two kinds of IOError:

1. An "expected" one. In this case, we indicate that we have encountered the EOF.
2. An "unexpected" one. In this case, we re-raise.

So, I would do something like:

```python
_OUT_OF_RANGE = 'Requested Range Not Satisfiable'
"""Returned by AWS when we try to seek beyond EOF."""

def _unwrap_ioerror(ioe):
    if ioe.__context__ and hasattr(ioe.__context__, 'response'):
        inner_error = ioe.__context__.response.get('Error')
        if inner_error.get('Message') == _OUT_OF_RANGE:
            return inner_error

try:
    ...
except IOError as ioe:
    inner_error = _unwrap_ioerror(ioe)
    if not inner_error:
        raise

    try:
        self._position = self._content_length = inner_error['ActualObjectSize']
    except KeyError:
        self._position = self._content_length = ... # handle moto quirk here
else:
    ...
```

Reply:
I made the helper function a little more general -- it returns the `'Error'` dict from the s3 response, which seems like it might be useful elsewhere at some point.